packages/plugin-evm/src/providers/wallet.ts

```diff
@@ -104,69 +63,81 @@ export class WalletProvider {
-        } catch (error: any) {
-            if (error.code === 4902) {
-                console.log(
-                    "[WalletProvider] Chain not added to wallet (error 4902) - attempting to add chain first"
-                );
-                await walletClient.addChain({
-                    chain: {
-                        ...getChainConfigs(runtime)[chain].chain,
-                        rpcUrls: {
-                            default: {
-                                http: [getChainConfigs(runtime)[chain].rpcUrl],
-                            },
-                            public: {
-                                http: [getChainConfigs(runtime)[chain].rpcUrl],
-                            },
-                        },
-                    },
-                });
-                await walletClient.switchChain({
-                    id: getChainConfigs(runtime)[chain].chainId,
-                });
-            } else {
-                throw error;
-            }
+            return formatUnits(balance, 18);
+        } catch (error) {
+            console.error("Error getting wallet balance:", error);
+            return null;
         }
-
-        this.currentChain = chain;
     }
 
-    getPublicClient(
-        chain: SupportedChain
-    ): PublicClient<HttpTransport, Chain, Account | undefined> {
-        return this.chainConfigs[chain].publicClient;
+    addChain(chain: SupportedChain) {
+        this.setChains([chain]);
     }
 
-    getWalletClient(): WalletClient {
-        const walletClient = this.chainConfigs[this.currentChain].walletClient;
-        if (!walletClient) throw new Error("Wallet not connected");
-        return walletClient;
+    switchChain(chain: SupportedChain) {
+        if (!this.chains[chain]) {
+            this.addChain(chain);
+        }
+        this.setCurrentChain(chain);
     }
 
-    getCurrentChain(): SupportedChain {
-        return this.currentChain;
-    }
+    getChainConfigs(chainName: SupportedChain): Chain {
+        const chain = viemChains[chainName];
 
-    getChainConfig(chain: SupportedChain) {
-        return getChainConfigs(this.runtime)[chain];
+        if (!chain?.id) {
+            throw new Error("Invalid chain name");
+        }
+
+        return chain;
     }
+
+    private setAccount = (pk: `0x${string}`) => {
+        this.account = privateKeyToAccount(pk);
+    };
+    private setChains = (chainNames?: SupportedChain[]) => {
+        if (!chainNames) {
+            return;
+        }
+        chainNames.forEach((name) => {
+            const chain = viemChains[name];
+
+            if (!chain?.id) {
+                throw new Error("Invalid chain name");
+            }
+
+            this.chains[name] = chain;
+        });
+    };
+    private setCurrentChain = (chain: SupportedChain) => {
+        this.currentChain = chain;
+    };
+    private createHttpTransport = (chain: SupportedChain) => {
+        return http(this.chains[chain].rpcUrls.default.http[0]);
```

Comment:
My concern with this is that the public RPCs can be quite unreliable. Here I'm not able to override the RPC with one I want to use (for instance, from Alchemy).

Maybe you can infer the RPC from the env variables through the chain name having the env variable be `ETHEREUM_PROVIDER_<CHAIN_NAME>` (e.g if I have base as a supported chain, I will look for an `ETHEREUM_PROVIDER_BASE` env variable to set up the RPC)

Reply:
@0xaguspunk totally agree, let me push some fixes in a bit

Reply:
@0xaguspunk This has been fixed in [commit ee6663e](https://github.com/ai16z/eliza/pull/864/commits/ee6663e917209bce13c58b55077763ed0141bbc3).

You can now configure the mainnet custom RPC URL by setting the EVM_PROVIDER_URL value. For other chains, use the template ETHEREUM_PROVIDER_<CHAIN_NAME>.

Additionally, the current version supports adding custom chains, even if they are not part of viem/chains.